src/form/bf_mx.jl

```diff
@@ -76,315 +43,748 @@ function variable_mc_branch_series_current_prod_hermitian(pm::AbstractUBFModels;
+function constraint_mc_model_voltage_magnitude_difference(pm::AbstractUBFModels, n::Int, i, f_bus, t_bus, f_idx, t_idx, r, x, g_sh_fr, b_sh_fr, tm)
+    Wr_fr = _PMs.var(pm, n, :Wr)[f_bus]
+    Wi_fr = _PMs.var(pm, n, :Wi)[f_bus]
 
-        p_mat_dict[i] = p_mat
-        q_mat_dict[i] = q_mat
-    end
-    _PMs.var(pm, nw)[:P_mx] = p_mat_dict
-    _PMs.var(pm, nw)[:Q_mx] = q_mat_dict
+    Wr_to = _PMs.var(pm, n, :Wr)[t_bus]
+    Wi_to = _PMs.var(pm, n, :Wi)[t_bus]
+
+    p_fr = _PMs.var(pm, n, :P)[f_idx]
+    q_fr = _PMs.var(pm, n, :Q)[f_idx]
+
+    p_s_fr = p_fr - (Wr_fr*(g_sh_fr)' + Wi_fr*(b_sh_fr)')
+    q_s_fr = q_fr - (Wi_fr*(g_sh_fr)' - Wr_fr*(b_sh_fr)')
+
+    CCr =  _PMs.var(pm, n, :CCr)[i]
+    CCi =  _PMs.var(pm, n, :CCi)[i]
+
+    #KVL over the line:
+    JuMP.@constraint(pm.model, diag(Wr_to) .== diag(
+    Wr_fr   - p_s_fr  *r' - q_s_fr*x'        - r*p_s_fr'    - x*q_s_fr'
+    + r*CCr*r' - x     *CCi*r' + x*CCr *x' + r*CCi *x'))
+    JuMP.@constraint(pm.model, _mat2utrivec!(Wr_to) .== _mat2utrivec!(
+    Wr_fr   - p_s_fr  *r' - q_s_fr*x'        - r*p_s_fr'    - x*q_s_fr'
+    + r*CCr*r' - x     *CCi*r' + x*CCr *x' + r*CCi *x'))
+    JuMP.@constraint(pm.model, _mat2utrivec!(Wi_to) .== _mat2utrivec!(
+    Wi_fr   - q_s_fr  *r' + p_s_fr*x'        - x*p_s_fr'    + r*q_s_fr'
+    + x*CCr*r' + r     *CCi*r' - r*CCr *x' + x*CCi *x'))
 end
 
 
-"variable: `p_lt[l,i,j]` for `(l,i,j)` in `arcs`"
-function variable_mc_lower_triangle_active_branch_flow(pm::_PMs.AbstractPowerModel; nw::Int=pm.cnw, cnd::Int=pm.ccnd, bounded = true)
-    smaxdict = Dict{Tuple{Int64, Int64, Int64}, Any}()
+"""
+For the matrix KCL formulation (mx), the generator needs an explicit current and
+power variable.
+"""
+function variable_mc_generation_mx(pm::AbstractUBFModels; nw=pm.cnw)
+    variable_mc_generation_current_mx(pm; nw=nw)
+    variable_mc_generation_power_mx(pm; nw=nw)
+end
 
-    for (l,i,j) in _PMs.ref(pm, nw, :arcs)
-        cmax = _PMs.ref(pm, nw, :branch, l, "rate_a").values./_PMs.ref(pm, nw, :bus, i, "vmin").values
-        smax = _PMs.ref(pm, nw, :bus, i, "vmax").values.*cmax'
-        smaxltri = _mat2ltrivec!(smax)
-        smaxdict[(l,i,j)] = smaxltri
-    end
 
-    if bounded
-        _PMs.var(pm, nw, cnd)[:p_lt] = JuMP.@variable(pm.model,
-        [(l,i,j) in _PMs.ref(pm, nw, :arcs)], base_name="$(nw)_$(cnd)_p_lt",
-        lower_bound = -smaxdict[(l,i,j)][cnd],
-        upper_bound =  smaxdict[(l,i,j)][cnd],
-        start = _PMs.comp_start_value(_PMs.ref(pm, nw, :branch, l), "p_start", cnd)
-        )
-    else
-        _PMs.var(pm, nw, cnd)[:p_lt] = JuMP.@variable(pm.model,
-        [(l,i,j) in _PMs.ref(pm, nw, :arcs)], base_name="$(nw)_$(cnd)_p_lt",
-        start = _PMs.comp_start_value(_PMs.ref(pm, nw, :branch, l), "p_start", cnd)
-        )
+"""
+For the matrix KCL formulation (mx), the generator needs an explicit power
+variable.
+"""
+function variable_mc_generation_power_mx(pm::AbstractUBFModels; nw=pm.cnw)
```

Comment:
maybe we can name this `variable_mx_generation_power` because it only for matrix formulations?

Reply:
Relates to same discussion below, will wait with this until that's cleared up.